File: JavaIntroNeuralNetworkEdition2/src/com/heatonresearch/book/introneuralnet/neural/feedforward/train/genetic/TrainingSetNeuralChromosome.java
/**
 * Introduction to Neural Networks with Java, 2nd Edition
 * Copyright 2008 by Heaton Research, Inc. 
 * http://www.heatonresearch.com/books/java-neural-2/
 * 
 * ISBN13: 978-1-60439-008-7  	 
 * ISBN:   1-60439-008-5
 *   
 * This class is released under the:
 * GNU Lesser General Public License (LGPL)
 * http://www.gnu.org/copyleft/lesser.html
 */
package com.heatonresearch.book.introneuralnet.neural.feedforward.train.genetic;

import com.heatonresearch.book.introneuralnet.neural.exception.NeuralNetworkError;
import com.heatonresearch.book.introneuralnet.neural.feedforward.FeedforwardNetwork;

/**
 * TrainingSetNeuralChromosome: Implements a chromosome that 
 * allows a feedforward neural network to be trained using a 
 * genetic algorithm.  The network is trained using training
 * sets.
 * 
 * The chromosome for a feed forward neural network
 * is the weight and threshold matrix.  
 * 
 * @author Jeff Heaton
 * @version 2.1
 */
public class TrainingSetNeuralChromosome extends
		NeuralChromosome<TrainingSetNeuralGeneticAlgorithm> {

	/**
	 * The constructor, takes a list of cities to set the initial "genes" to.
	 * 
	 * @param cities
	 *            The order that this chromosome would visit the cities. These
	 *            cities can be thought of as the genes of this chromosome.
	 * @throws NeuralNetworkException
	 */
	public TrainingSetNeuralChromosome(
			final TrainingSetNeuralGeneticAlgorithm genetic,
			final FeedforwardNetwork network) throws NeuralNetworkError {
		this.setGeneticAlgorithm(genetic);
		this.setNetwork(network);

		initGenes(network.getWeightMatrixSize());
		updateGenes();
	}

	@Override
	public void calculateCost() throws NeuralNetworkError {
		// update the network with the new gene values
		this.updateNetwork();

		// update the cost with the new genes
		final double input[][] = this.getGeneticAlgorithm().getInput();
		final double ideal[][] = this.getGeneticAlgorithm().getIdeal();

		setCost(getNetwork().calculateError(input, ideal));

	}

	/**
	 * Set all genes.
	 * 
	 * @param list
	 *            A list of genes.
	 * @throws NeuralNetworkException
	 */
	@Override
	public void setGenes(final Double[] list) throws NeuralNetworkError {

		// copy the new genes
		super.setGenes(list);

		calculateCost();

	}
}

File: CSIntroNeuralNetworkEdition2/HeatonResearchNeural/Feedforward/Train/Genetic/TrainingSetNeuralChromosome.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using HeatonResearchNeural.Genetic;

namespace HeatonResearchNeural.Feedforward.Train.Genetic
{
    class TrainingSetNeuralChromosome : NeuralChromosome
    {
        /// <summary>
        /// The genes that make up this chromosome.
        /// </summary>
        public override double[] Genes
        {
            get
            {
                return base.Genes;
            }

            set
            {
                // copy the new genes
                base.Genes = value;

                CalculateCost();
            }
        }


        /// <summary>
        /// The constructor sets up to train the neural network with a GA.
        /// </summary>
        /// <param name="genetic">The genetic algorithm to use.</param>
        /// <param name="network">The neural network to use.</param>
        public TrainingSetNeuralChromosome(
                 TrainingSetNeuralGeneticAlgorithm genetic,
                 FeedforwardNetwork network)
        {
            this.GA =  genetic;
            this.Network = network;

            InitGenes(network.MatrixSize);
            UpdateGenes();
        }

        /// <summary>
        /// Calculate the cost for this chromosome.
        /// </summary>
        override public void CalculateCost()
        {
            // update the network with the new gene values
            this.UpdateNetwork();

            // update the cost with the new genes
            double[][] input = this.getTrainingSetNeuralGeneticAlgorithm().Input;
            double[][] ideal = this.getTrainingSetNeuralGeneticAlgorithm().Ideal;

            this.Cost  = this.Network.CalculateError(input, ideal);

        }

        /// <summary>
        /// Get the genetic algorithm used for this chromosome.
        /// </summary>
        /// <returns>The genetic algorithm used with this chromosome.</returns>
        public TrainingSetNeuralGeneticAlgorithm getTrainingSetNeuralGeneticAlgorithm()
        {
            return (TrainingSetNeuralGeneticAlgorithm)this.GA;
        }
    }
}
